caffe2/python/ideep/transform_ideep_net.py

```diff
@@ -0,0 +1,302 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from caffe2.proto import caffe2_pb2
+from caffe2.python import core, workspace, utils
+import caffe2.python._import_c_extension as C
+
+import numpy as np
+import argparse
+import copy
+import os.path
+
+
+def pairwise(iterable):
+    from itertools import tee
+    a, b = tee(iterable)
+    next(b, None)
+    return zip(a, b)
+
+
+def last_producer(ops, blob):
+    for (i, op) in reversed(list(enumerate(ops))):
+        if blob in op.output:
+            return i
+    raise ValueError("Failed to find last producer of blob, %s", blob)
+
+
+def blob_uses(net, blob):
+    u = []
+    for i, op in enumerate(net.op):
+        if blob in op.input or blob in op.control_input:
+            u.append(i)
+    return u
+
+
+def GetArgumentParser():
+    parser = argparse.ArgumentParser(description="Caffe2 optimization")
+    parser.add_argument("--init_net", type=str, help="init net")
+    parser.add_argument("--pred_net", type=str, help="predict net")
+    parser.add_argument("--fuse_bn", default=False, action='store_true')
+    parser.add_argument("--fuse_mul_add", default=False, action='store_true')
+    parser.add_argument("--fuse_conv_relu", default=False, action='store_true')
+    return parser
+
+
+def fuse_first_bn(net, params, removed_tensors):
+    net = copy.deepcopy(net)
+    params = copy.deepcopy(params)
+
+    for ((i, current), (j, next_)) in pairwise(enumerate(net.op)):
+        if next_.input[0] != current.output[0]:
```

Comment:
this can be improved to based on "next usage" instead of physically "next position"